File: src/routes/lines.ts
import { Hono } from 'hono'
import { streamSSE } from 'hono/streaming'
import { eq, and, gt, max, sql } from 'drizzle-orm'
import { db } from '../db/index.js'
import { projects, lines } from '../db/schema.js'
import { ScriptError, validateScript } from '../script.js'
import { analyzeText } from '../analyze.js'
import { serializeLine } from './serialize.js'

const app = new Hono()

const nowIso = () => new Date().toISOString()

function touchProject(projectId: number) {
  db.update(projects).set({ updated_at: nowIso() }).where(eq(projects.id, projectId)).run()
}

function nextPosition(projectId: number): number {
  const row = db
    .select({ value: max(lines.position) })
    .from(lines)
    .where(eq(lines.project_id, projectId))
    .get()
  return (row?.value ?? -1) + 1
}

// script (配列) を検証して保存用文字列にする。不正なら ScriptError を投げる。
function serializeScript(input: unknown): string {
  return JSON.stringify(validateScript(input))
}

// 行追加 (position は末尾)
app.post('/api/projects/:id/lines', async (c) => {
  const projectId = Number(c.req.param('id'))
  if (Number.isNaN(projectId)) return c.json({ error: 'invalid id' }, 400)

  const project = db.select().from(projects).where(eq(projects.id, projectId)).get()
  if (!project) return c.json({ error: 'not found' }, 404)

  const body = await c.req.json().catch(() => ({}))
  const mode = body.mode === 'acting' ? 'acting' : 'announcer'
  const text = typeof body.text === 'string' ? body.text.trim() : ''
  if (!text) return c.json({ error: 'text required' }, 400)

  let script: string | null = null
  if (mode === 'acting' && body.script != null) {
    try {
      script = serializeScript(body.script)
    } catch (e) {
      return c.json({ error: e instanceof ScriptError ? e.message : 'invalid script' }, 400)
    }
  }

  const row = db
    .insert(lines)
    .values({ project_id: projectId, position: nextPosition(projectId), mode, text, script })
    .returning()
    .get()
  touchProject(projectId)
  return c.json(serializeLine(row), 201)
})

// 行更新 (text / script / mode)。script は validateScript を通してから保存 (不正は 400)。
app.patch('/api/lines/:id', async (c) => {
  const id = Number(c.req.param('id'))
  if (Number.isNaN(id)) return c.json({ error: 'invalid id' }, 400)

  const line = db.select().from(lines).where(eq(lines.id, id)).get()
  if (!line) return c.json({ error: 'not found' }, 404)

  const body = await c.req.json().catch(() => ({}))
  const patch: Partial<typeof lines.$inferInsert> = {}

  if (typeof body.text === 'string') {
    const t = body.text.trim()
    if (!t) return c.json({ error: 'text must not be empty' }, 400)
    patch.text = t
  }
  if (body.mode === 'announcer' || body.mode === 'acting') {
    patch.mode = body.mode
  }
  if ('script' in body) {
    if (body.script == null) {
      patch.script = null
    } else {
      try {
        patch.script = serializeScript(body.script)
      } catch (e) {
        return c.json({ error: e instanceof ScriptError ? e.message : 'invalid script' }, 400)
      }
    }
  }

  const row = db.update(lines).set(patch).where(eq(lines.id, id)).returning().get()
  touchProject(line.project_id)
  return c.json(serializeLine(row))
})

// 行複製。mode/text/script ごとコピーして対象行の直下 (position+1) に挿入する。
// 感情パラメータの A/B 比較用。挿入前に後続 position を +1 シフトする。
app.post('/api/lines/:id/duplicate', (c) => {
  const id = Number(c.req.param('id'))
  if (Number.isNaN(id)) return c.json({ error: 'invalid id' }, 400)

  const line = db.select().from(lines).where(eq(lines.id, id)).get()
  if (!line) return c.json({ error: 'not found' }, 404)

  const row = db.transaction((tx) => {
    tx.update(lines)
      .set({ position: sql`${lines.position} + 1` })
      .where(and(eq(lines.project_id, line.project_id), gt(lines.position, line.position)))
      .run()
    return tx
      .insert(lines)
      .values({
        project_id: line.project_id,
        position: line.position + 1,
        mode: line.mode,
        text: line.text,
        script: line.script,
      })
      .returning()
      .get()
  })
  touchProject(line.project_id)
  return c.json(serializeLine(row), 201)
})

// 行削除
app.delete('/api/lines/:id', (c) => {
  const id = Number(c.req.param('id'))
  if (Number.isNaN(id)) return c.json({ error: 'invalid id' }, 400)

  const line = db.select().from(lines).where(eq(lines.id, id)).get()
  if (!line) return c.json({ error: 'not found' }, 404)

  db.delete(lines).where(eq(lines.id, id)).run()
  touchProject(line.project_id)
  return c.json({ ok: true })
})

// 並び順一括更新 (id 配列を受け position を 0..n で振り直す)
app.put('/api/projects/:id/lines/order', async (c) => {
  const projectId = Number(c.req.param('id'))
  if (Number.isNaN(projectId)) return c.json({ error: 'invalid id' }, 400)

  const project = db.select().from(projects).where(eq(projects.id, projectId)).get()
  if (!project) return c.json({ error: 'not found' }, 404)

  const body = await c.req.json().catch(() => ({}))
  const order = Array.isArray(body.order) ? body.order.map(Number) : null
  if (!order || order.some(Number.isNaN)) return c.json({ error: 'order must be an array of ids' }, 400)

  const existing = db.select({ id: lines.id }).from(lines).where(eq(lines.project_id, projectId)).all()
  const existingIds = new Set(existing.map((r) => r.id))
  if (order.length !== existingIds.size || order.some((oid: number) => !existingIds.has(oid))) {
    return c.json({ error: 'order must contain exactly the ids of this project' }, 400)
  }

  db.transaction((tx) => {
    order.forEach((lineId: number, index: number) => {
      tx.update(lines).set({ position: index }).where(eq(lines.id, lineId)).run()
    })
  })
  touchProject(projectId)

  const rows = db.select().from(lines).where(eq(lines.project_id, projectId)).orderBy(lines.position).all()
  return c.json(rows.map(serializeLine))
})

// テキスト流し込み。改行で分割し空行は無視。
//   announcer: 全行を一括 insert して JSON で即返す。
//   acting:    1行ずつ analyze しながら SSE で進捗を push (失敗行は announcer として保存)。
app.post('/api/projects/:id/import', async (c) => {
  const projectId = Number(c.req.param('id'))
  if (Number.isNaN(projectId)) return c.json({ error: 'invalid id' }, 400)

  const project = db.select().from(projects).where(eq(projects.id, projectId)).get()
  if (!project) return c.json({ error: 'not found' }, 404)

  const body = await c.req.json().catch(() => ({}))
  const mode = body.mode === 'acting' ? 'acting' : 'announcer'
  const texts = String(body.text ?? '')
    .split(/\r?\n/)
    .map((s) => s.trim())
    .filter(Boolean)

  // `after` (行id) 指定時はその行の直後に挿入する。指定が無い / 他プロジェクトの行なら
  // 現行どおり末尾追加 (完全後方互換)。挿入本数ぶん後続 position を +1 シフトしてから
  // startPos = after行.position + 1 で詰める。
  let startPos = nextPosition(projectId)
  if (body.after != null && texts.length > 0) {
    const afterId = Number(body.after)
    const afterRow = Number.isNaN(afterId)
      ? undefined
      : db.select().from(lines).where(eq(lines.id, afterId)).get()
    if (afterRow && afterRow.project_id === projectId) {
      db.update(lines)
        .set({ position: sql`${lines.position} + ${texts.length}` })
        .where(and(eq(lines.project_id, projectId), gt(lines.position, afterRow.position)))
        .run()
      startPos = afterRow.position + 1
    }
  }

  if (mode === 'announcer') {
    if (texts.length === 0) return c.json({ mode, created: [] })
    const created = db
      .insert(lines)
      .values(texts.map((text, i) => ({
        project_id: projectId,
        position: startPos + i,
        mode: 'announcer' as const,
        text,
        script: null,
      })))
      .returning()
      .all()
    touchProject(projectId)
    return c.json({ mode, created: created.map(serializeLine) })
  }

  // acting: SSE で 1行ずつ進捗を流す
  const analyzeCmd = c.get('analyzeCmd')
  return streamSSE(c, async (stream) => {
    const total = texts.length
    for (let i = 0; i < total; i++) {
      if (stream.aborted) break // client disconnected — saved rows remain
      const text = texts[i]
      let lineMode: 'announcer' | 'acting' = 'acting'
      let script: string | null = null
      try {
        script = JSON.stringify(await analyzeText(text, analyzeCmd))
      } catch {
        // 分析失敗は announcer 行として保存し、データを失わない
        lineMode = 'announcer'
        script = null
      }
      const row = db
        .insert(lines)
        .values({ project_id: projectId, position: startPos + i, mode: lineMode, text, script })
        .returning()
        .get()
      await stream.writeSSE({
        data: JSON.stringify({
          index: i + 1,
          total,
          status: lineMode === 'acting' ? 'done' : 'failed',
          line: serializeLine(row),
        }),
      })
    }
    if (total > 0) touchProject(projectId)
    await stream.writeSSE({ event: 'complete', data: JSON.stringify({ total }) })
  })
})

export default app

File: src/routes/lines.rs
// 行 (lines) CRUD + 並べ替え + 流し込み。TS 版 src-ts/routes/lines.ts を契約互換に移植。
// import の acting モードは analyzer (R3) 依存のため 501 を返す。

use crate::db::now_iso;
use crate::error::AppError;
use crate::script::serialize_script;
use crate::serialize::{get_line, lines_by_project, serialize_line, LineRow};
use crate::state::AppState;
use axum::body::Bytes;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::routing::{patch, post, put};
use axum::{Json, Router};
use rusqlite::{Connection, OptionalExtension};
use serde_json::{json, Value};

use super::{parse_body, parse_id, JsonResult};

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/api/projects/{id}/lines", post(add))
        .route("/api/lines/{id}", patch(update).delete(remove))
        .route("/api/lines/{id}/duplicate", post(duplicate))
        .route("/api/projects/{id}/lines/order", put(reorder))
        .route("/api/projects/{id}/import", post(import))
}

fn touch_project(conn: &Connection, project_id: i64) -> rusqlite::Result<()> {
    conn.execute(
        "UPDATE projects SET updated_at = ?1 WHERE id = ?2",
        (now_iso(), project_id),
    )?;
    Ok(())
}

fn next_position(conn: &Connection, project_id: i64) -> rusqlite::Result<i64> {
    let max: Option<i64> = conn.query_row(
        "SELECT max(position) FROM lines WHERE project_id = ?1",
        [project_id],
        |r| r.get(0),
    )?;
    Ok(max.unwrap_or(-1) + 1)
}

fn project_exists(conn: &Connection, id: i64) -> rusqlite::Result<bool> {
    let found: Option<i64> = conn
        .query_row("SELECT id FROM projects WHERE id = ?1", [id], |r| r.get(0))
        .optional()?;
    Ok(found.is_some())
}

// 行追加 (position は末尾)
async fn add(
    State(state): State<AppState>,
    Path(id): Path<String>,
    body: Bytes,
) -> JsonResult {
    let project_id = parse_id(&id)?;
    let body = parse_body(&body);

    let mode = if body.get("mode").and_then(Value::as_str) == Some("acting") {
        "acting"
    } else {
        "announcer"
    };
    let text = body
        .get("text")
        .and_then(Value::as_str)
        .map(str::trim)
        .unwrap_or("");

    let conn = state.db.lock().unwrap();
    if !project_exists(&conn, project_id)? {
        return Err(AppError::NotFound("not found".into()));
    }
    if text.is_empty() {
        return Err(AppError::BadRequest("text required".into()));
    }

    // script は mode==acting かつ body.script が非 null の時のみ検証・保存する。
    let mut script: Option<String> = None;
    if mode == "acting" {
        if let Some(s) = body.get("script") {
            if !s.is_null() {
                script = Some(serialize_script(s)?);
            }
        }
    }

    let position = next_position(&conn, project_id)?;
    let row = conn.query_row(
        "INSERT INTO lines (project_id, position, mode, text, script) \
         VALUES (?1, ?2, ?3, ?4, ?5) RETURNING id, project_id, position, mode, text, script",
        (project_id, position, mode, text, &script),
        LineRow::from_row,
    )?;
    touch_project(&conn, project_id)?;
    Ok((StatusCode::CREATED, Json(serialize_line(&row))))
}

// 行更新 (text / script / mode)。script は validateScript を通してから保存 (不正は 400)。
async fn update(
    State(state): State<AppState>,
    Path(id): Path<String>,
    body: Bytes,
) -> JsonResult {
    let id = parse_id(&id)?;
    let body = parse_body(&body);

    let conn = state.db.lock().unwrap();
    let Some(line) = get_line(&conn, id)? else {
        return Err(AppError::NotFound("not found".into()));
    };

    let mut new_text = line.text.clone();
    if let Some(t) = body.get("text").and_then(Value::as_str) {
        let t = t.trim();
        if t.is_empty() {
            return Err(AppError::BadRequest("text must not be empty".into()));
        }
        new_text = t.to_string();
    }

    let mut new_mode = line.mode.clone();
    match body.get("mode").and_then(Value::as_str) {
        Some(m @ "announcer") | Some(m @ "acting") => new_mode = m.to_string(),
        _ => {}
    }

    // script は「キーが有る&null」「有る&値」「無い」の3状態を区別する。
    let mut new_script = line.script.clone();
    if let Some(s) = body.get("script") {
        if s.is_null() {
            new_script = None;
        } else {
            new_script = Some(serialize_script(s)?);
        }
    }

    let row = conn.query_row(
        "UPDATE lines SET text = ?1, mode = ?2, script = ?3 WHERE id = ?4 \
         RETURNING id, project_id, position, mode, text, script",
        (&new_text, &new_mode, &new_script, id),
        LineRow::from_row,
    )?;
    touch_project(&conn, line.project_id)?;
    Ok((StatusCode::OK, Json(serialize_line(&row))))
}

// 行複製。mode/text/script ごとコピーして対象行の直下 (position+1) に挿入する。
async fn duplicate(State(state): State<AppState>, Path(id): Path<String>) -> JsonResult {
    let id = parse_id(&id)?;
    let mut conn = state.db.lock().unwrap();
    let Some(line) = get_line(&conn, id)? else {
        return Err(AppError::NotFound("not found".into()));
    };

    let row = {
        let tx = conn.transaction()?;
        tx.execute(
            "UPDATE lines SET position = position + 1 WHERE project_id = ?1 AND position > ?2",
            (line.project_id, line.position),
        )?;
        let row = tx.query_row(
            "INSERT INTO lines (project_id, position, mode, text, script) \
             VALUES (?1, ?2, ?3, ?4, ?5) RETURNING id, project_id, position, mode, text, script",
            (
                line.project_id,
                line.position + 1,
                &line.mode,
                &line.text,
                &line.script,
            ),
            LineRow::from_row,
        )?;
        tx.commit()?;
        row
    };
    touch_project(&conn, line.project_id)?;
    Ok((StatusCode::CREATED, Json(serialize_line(&row))))
}

// 行削除
async fn remove(State(state): State<AppState>, Path(id): Path<String>) -> JsonResult {
    let id = parse_id(&id)?;
    let conn = state.db.lock().unwrap();
    let Some(line) = get_line(&conn, id)? else {
        return Err(AppError::NotFound("not found".into()));
    };
    conn.execute("DELETE FROM lines WHERE id = ?1", [id])?;
    touch_project(&conn, line.project_id)?;
    Ok((StatusCode::OK, Json(json!({ "ok": true }))))
}

// JS の Number(value) 相当。数値化できなければ None (NaN)。
fn to_number(value: &Value) -> Option<f64> {
    match value {
        Value::Number(n) => n.as_f64(),
        Value::String(s) => {
            let t = s.trim();
            if t.is_empty() {
                Some(0.0)
            } else {
                t.parse::<f64>().ok()
            }
        }
        _ => None,
    }
}

// 並び順一括更新 (id 配列を受け position を 0..n で振り直す)
async fn reorder(
    State(state): State<AppState>,
    Path(id): Path<String>,
    body: Bytes,
) -> JsonResult {
    let project_id = parse_id(&id)?;
    let body = parse_body(&body);

    let mut conn = state.db.lock().unwrap();
    if !project_exists(&conn, project_id)? {
        return Err(AppError::NotFound("not found".into()));
    }

    let Some(arr) = body.get("order").and_then(Value::as_array) else {
        return Err(AppError::BadRequest("order must be an array of ids".into()));
    };
    let mut order: Vec<i64> = Vec::with_capacity(arr.len());
    for v in arr {
        match to_number(v) {
            Some(n) => order.push(n as i64),
            None => return Err(AppError::BadRequest("order must be an array of ids".into())),
        }
    }

    let existing_set: std::collections::HashSet<i64> = {
        let mut stmt = conn.prepare("SELECT id FROM lines WHERE project_id = ?1")?;
        let rows = stmt.query_map([project_id], |r| r.get(0))?;
        rows.collect::<rusqlite::Result<_>>()?
    };
    if order.len() != existing_set.len() || order.iter().any(|oid| !existing_set.contains(oid)) {
        return Err(AppError::BadRequest(
            "order must contain exactly the ids of this project".into(),
        ));
    }

    {
        let tx = conn.transaction()?;
        for (index, line_id) in order.iter().enumerate() {
            tx.execute(
                "UPDATE lines SET position = ?1 WHERE id = ?2",
                (index as i64, line_id),
            )?;
        }
        tx.commit()?;
    }
    touch_project(&conn, project_id)?;

    let rows: Vec<Value> = lines_by_project(&conn, project_id)?
        .iter()
        .map(serialize_line)
        .collect();
    Ok((StatusCode::OK, Json(Value::Array(rows))))
}

// テキスト流し込み。改行で分割し空行は無視。announcer は一括 insert して即返す。
// acting は R3 (analyzer/SSE) 依存のため 501 を返す。
async fn import(
    State(state): State<AppState>,
    Path(id): Path<String>,
    body: Bytes,
) -> JsonResult {
    let project_id = parse_id(&id)?;
    let body = parse_body(&body);

    let conn = state.db.lock().unwrap();
    if !project_exists(&conn, project_id)? {
        return Err(AppError::NotFound("not found".into()));
    }

    if body.get("mode").and_then(Value::as_str) == Some("acting") {
        return Err(AppError::NotImplemented(
            "acting import not implemented".into(),
        ));
    }

    let raw = body.get("text").and_then(Value::as_str).unwrap_or("");
    // TS: split(/\r?\n/).map(trim).filter(Boolean)。trim() は末尾 \r も除く。
    let texts: Vec<&str> = raw
        .split('\n')
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .collect();

    let mut start_pos = next_position(&conn, project_id)?;
    if !texts.is_empty() {
        if let Some(after) = body.get("after") {
            if !after.is_null() {
                if let Some(after_id) = to_number(after).map(|n| n as i64) {
                    let after_row = get_line(&conn, after_id)?;
                    if let Some(after_row) = after_row {
                        if after_row.project_id == project_id {
                            conn.execute(
                                "UPDATE lines SET position = position + ?1 \
                                 WHERE project_id = ?2 AND position > ?3",
                                (texts.len() as i64, project_id, after_row.position),
                            )?;
                            start_pos = after_row.position + 1;
                        }
                    }
                }
            }
        }
    }

    if texts.is_empty() {
        return Ok((
            StatusCode::OK,
            Json(json!({ "mode": "announcer", "created": [] })),
        ));
    }

    let mut created: Vec<Value> = Vec::with_capacity(texts.len());
    for (i, text) in texts.iter().enumerate() {
        let row = conn.query_row(
            "INSERT INTO lines (project_id, position, mode, text, script) \
             VALUES (?1, ?2, 'announcer', ?3, NULL) \
             RETURNING id, project_id, position, mode, text, script",
            (project_id, start_pos + i as i64, text),
            LineRow::from_row,
        )?;
        created.push(serialize_line(&row));
    }
    touch_project(&conn, project_id)?;
    Ok((
        StatusCode::OK,
        Json(json!({ "mode": "announcer", "created": created })),
    ))
}
